src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.Analyzers.targets

```diff
@@ -77,6 +77,13 @@ Copyright (c) .NET Foundation. All rights reserved.
     <WarningLevel Condition="'$(Language)' == 'C#'">6</WarningLevel>
   </PropertyGroup>
 
+  <!-- Set the default WarningLevel based on EffectiveAnalysisLevel
+       For .NET 7 we want the Warning level to be 7 -->
+  <PropertyGroup Condition="'$(EffectiveAnalysisLevel)' != '' And
+                             $([MSBuild]::VersionGreaterThanOrEquals($(EffectiveAnalysisLevel), '7.0'))">
+    <WarningLevel Condition="'$(Language)' == 'C#'">7</WarningLevel>
```

Comment:
I'm curious if this can override a value the user has explicitly set for `<WarningLevel>`?

Should the condition be `'$(Language)' == 'C#' And '$(WarningLevel)' == ''`?

Reply:
It seems that `WarningLevel` isn't intended to be the user-facing field, instead `AnalysisLevel` is that field. We do similar defaulting for that property [here](https://github.com/dotnet/sdk/blob/bc24f09e31688ae3013a3150d8cbdb2c120bb236/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.Analyzers.targets#L15-L51), but it too is missing defaults/values for 7.0 (which I can add here).

The intent of the two different-but-related properties isn't entirely clear to me, I'll check our docs on docs.ms.com and see what we recommend to try and glean more context.

Reply:
@baronfel I think it *is* user-facing. It's documented [here](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/errors-warnings#warninglevel)

I think most properties that are not intended to be user-facing are prefixed with an underscore.

Reply:
Of course you're correct on both counts - I had only found the AnalysisLevel documentation (which exists at the SDK docs, rather than the compiler docs)

Reply:
> new WarningLevel for .Net 7

I'm not sure what you mean by this.

---

> separately investigating the way we've tied WarningLevel and AnalysisLevel together

It's okay to have WarningLevel tied with AnalysisLevel. Only have a question: if the user sets both AnalysisLevel and WarningLevel, which value of WarningLevel will take effect (ie, passed to the compiler)? The one the user set in his csproj? or the one from the SDK [here](https://github.com/dotnet/sdk/blob/bc24f09e31688ae3013a3150d8cbdb2c120bb236/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.Analyzers.targets#L82-L85)? If the one from the SDK can take effect over the user's defined one, then we need to adjust the condition to add `And '$(WarningLevel)' != ''`

---

> because you assert that we basically always want the latest WarningLevel regardless of user-specified AnalysisLevel

This will break backward compatibility badly, hence, beats the purpose of warning waves feature.

What I meant is adding `<WarningLevel>7</WarningLevel>` in .NET 7 templates.

---

Adding @jmarolf @mavasani for thoughts too.

Reply:
To summarize, there are two points here:


1. I confirmed that AnalysisLevel modifies WarningLevel in a very bad way. Given the following csproj:

    ```xml
    <Project Sdk="Microsoft.NET.Sdk">

      <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>net7.0</TargetFramework>
        <ImplicitUsings>enable</ImplicitUsings>
        <Nullable>enable</Nullable>
        <WarningLevel>7</WarningLevel>
        <AnalysisLevel>5</AnalysisLevel>
      </PropertyGroup>

    </Project>
    ```

    I see the following in binlog:

    ![image](https://user-images.githubusercontent.com/31348972/163420823-38d42c9a-b0ce-4084-8d85-69b91816a87c.png)

    I'm **not** opposed to the SDK setting WarningLevel, but it should do that only if the user doesn't explicitly provide a value for it.

    I'll consider this as a bug.

---

2. This one is a feature request, to let the default template have all the new compiler warnings that are introduced in a warning wave. It's simply adding `<WarningLevel>` in csproj in the templates.

Reply:
I completely agree with (1) - we should only ever conditionally-set WarningLevel in the SDK.
I think we can implement (2) purely by the mechanism already here in the SDK - if no user-set WarningLevel is present, default to the level for the TFM chosen (so net6.0 -> 6, net7.0 -> 7, and so on).

I've pushed up a version of this with tests to illustrate the behavior now, can you see if you agree?

My main concern is crowding the project templates, we pretty aggressively try to limit the content generated in them. I think we can get the auto-incrementing behavior here and get to the more-safe-by-default you're asking for.

Reply:
> if no user-set WarningLevel is present, default to the level for the TFM chosen (so net6.0 -> 6, net7.0 -> 7, and so on).

I don't disagree here. Seems fine, but needs attention/approval from Roslyn team I think.

Reply:
I've roped them in on an email for other parts of this (see my review comments below) so we should get alignment on it one way or the other.